pkg/registry/graph.go

```diff
@@ -0,0 +1,176 @@
+package registry
+
+import (
+	"fmt"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// Component contains the name of a registry component and pointers for
+// either a reference, chain, or workflow
+type Component struct {
+	Name      string
+	Reference *api.LiteralTestStep
+	Chain     []api.TestStep
+	Workflow  *api.MultiStageTestConfiguration
+}
+
+type nodeSet map[*Node]sets.Empty
+
+// Node contains a registry component, a set of its parents, and a set of its children
+type Node struct {
+	data     Component
+	parents  nodeSet
+	children nodeSet
+}
+
+// GetComponent return the registry component stored in the node
+func (n *Node) GetComponent() Component {
+	return n.data
+}
+
+// NodeByName is a mapping of component name to the graph node for the component
+type NodeByName map[string]*Node
+
+func newNode(component Component) *Node {
+	return &Node{
+		data:     component,
+		parents:  make(nodeSet),
+		children: make(nodeSet),
+	}
+}
+
+func (set nodeSet) insert(node *Node) {
+	set[node] = sets.Empty{}
+}
+
+func (set nodeSet) list() []*Node {
+	res := make([]*Node, 0, len(set))
+	for key := range set {
+		res = append(res, key)
+	}
+	return res
+}
+
+// addChild adds the provided node to the current node's set of children
+// and adds the current node to the provided node's set of parents
+func (n *Node) addChild(child *Node) {
+	n.children.insert(child)
+	child.parents.insert(n)
+}
+
+// AddAncestorNames gets the set of all ancestors for the node
+func (n *Node) GetAncestorNames() sets.String {
+	ancestors := sets.NewString()
+	for parent := range n.parents {
+		ancestors.Insert(parent.data.Name)
+		ancestors.Insert(parent.GetAncestorNames().List()...)
+	}
+	return ancestors
+}
+
+// GetDescendantNames gets the set of all descendants for the node
+func (n *Node) GetDescendantNames() sets.String {
+	descendants := sets.NewString()
+	for child := range n.children {
+		descendants.Insert(child.data.Name)
+		descendants.Insert(child.GetDescendantNames().List()...)
+	}
+	return descendants
+}
+
+func hasCycles(root *Node, ancestors sets.String, traversedPath []string) error {
+	if ancestors.Has(root.data.Name) {
+		return fmt.Errorf("Cycle detected: %s is an ancestor of itself; traversedPath: %v", root.data.Name, traversedPath)
+	}
+	ancestors.Insert(root.data.Name)
+	for child := range root.children {
+		// get new copy of ancestors and traversedPath so the root node's set isn't changed
+		ancestorsCopy := sets.NewString()
+		for _, ancestor := range ancestors.List() {
+			ancestorsCopy.Insert(ancestor)
+		}
+		traversedPathCopy := append(traversedPath[:0:0], traversedPath...)
+		if err := hasCycles(child, ancestorsCopy, traversedPathCopy); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// HasCycles detects any cycles in the descendants of the provided root node
+func HasCycles(root *Node) error {
+	return hasCycles(root, sets.NewString(), []string{})
```

Comment:
This assumes a tree, not a forest

Reply:
In `NewGraph`, a `root` node with an empty name field is created whose children are all nodes that don't have any parents after everything is resolved. If we pass the `root` node to `HasCycles`, it should be able to traverse the entire graph and find all cycles. Another option that would work would be to pass the `NodesByName` type, but either way should work.

Reply:
Why should we prefer a synthetic root with no content versus allowing a forest?

Reply:
It doesn't really matter, that's just how I first thought to do it (I've worked with trees a lot more than graphs/forests so this made sense to me). We could just not have a single root and use only the `NodesByName` map to access the graph's contents.

Reply:
IMO synthetic valueless roots for forests often get mistaken by someone who didn't write the implementation in the first place so I'd suggest not doing that.